### PR TITLE
Make mode an argument of goto convert methods

### PR DESCRIPTION
**Pull request by romainbrenguier**
This passes the mode as an argument of these functions instead of trying
to deduce it from the expression.
This avoids creating symbols with empty mode which can lead to problems
in later stages of the analysis.

### CODE REVIEW

File: src/goto-programs/goto_convert.cpp
Comment (tautschnig) [inline src/goto-programs/goto_convert.cpp]: Nit pick: could we use `ns.lookup(i.function).mode` here?

File: src/goto-programs/goto_convert.cpp
Comment (tautschnig) [inline src/goto-programs/goto_convert.cpp]: Should be `mode`, not `ID_java`.

File: src/goto-programs/goto_convert.cpp
Comment (tautschnig) [inline src/goto-programs/goto_convert.cpp]: Should be `mode`, not `ID_java`.

File: src/goto-programs/goto_convert.cpp
Comment (tautschnig) [inline src/goto-programs/goto_convert.cpp]: Should be `mode`, not `ID_java`.

File: src/goto-programs/goto_convert.cpp
Comment (tautschnig) [inline src/goto-programs/goto_convert.cpp]: Should be `mode`, not `ID_java`.

File: src/goto-programs/goto_convert.cpp
Comment (tautschnig) [inline src/goto-programs/goto_convert.cpp]: Should be `mode`, not `ID_java`.

File: src/goto-programs/goto_convert.cpp
Comment (tautschnig) [inline src/goto-programs/goto_convert.cpp]: Should be `mode`, not `ID_java`.

File: src/goto-programs/goto_convert.cpp
Comment (tautschnig) [inline src/goto-programs/goto_convert.cpp]: Must use `mode`, not `expr.get(ID_mode)`

File: src/goto-programs/goto_convert_side_effect.cpp
Comment (tautschnig) [inline src/goto-programs/goto_convert_side_effect.cpp]: Should be `mode`, not `ID_java`.

File: src/goto-programs/goto_convert_side_effect.cpp
Comment (tautschnig) [inline src/goto-programs/goto_convert_side_effect.cpp]: This one isn't necessary, `temporary_object` actually does have a mode provided.

**Review by tautschnig — CHANGES_REQUESTED** [review]
Thank you for this large-scale rework! A few `ID_java` slipped in/slipped through, but otherwise this looks like the set-up that I expected. Please do add "Fixes: #2138" to some commit message.

File: src/goto-programs/goto_convert.cpp
Comment (thk123) [inline src/goto-programs/goto_convert.cpp:2108]: Reluctant to suggest this now: but to avoid having this checked at run time - consider making mode an enum that we store in the irep using just index?

File: src/goto-programs/goto_convert.cpp
Comment (thk123) [reply src/goto-programs/goto_convert.cpp]: If you end up leaving as it is, use `lookup_ref` since immediately dereferencing

File: src/goto-programs/goto_convert_class.h
Comment (thk123) [inline src/goto-programs/goto_convert_class.h:54]: I wonder if easier to set up a `goto_convertert` with a mode rather than adding an extra param everywhere

**Review by thk123 — COMMENTED** [review]
Needs a TG pointer bump unless I've missed it. Couple of large scale suggestions which I'm happy to accept as out of scope.

File: src/goto-programs/goto_convert_class.h
Comment (romainbrenguier) [reply src/goto-programs/goto_convert_class.h:54]: Having it as a leaves the possibility to use the same converter for functions of different languages, and could make it easier to make some functions static. Now that the changes are done, I would say that it is easier to take that solution.

**Comment by romainbrenguier** [comment]
@tautschnig thanks for spotting the mistakes
@thk123 TG update is this https://github.com/diffblue/test-gen/pull/1762

File: src/goto-programs/goto_convert.cpp
Comment (romainbrenguier) [reply src/goto-programs/goto_convert.cpp]: Changing this to `ns.lookup(i.function).mode` or to `ns.lookup_ref(i.function).mode` make most of the tests in cbmc-java fail (with an identifier not found exception in the first case and core dumped in the second case)

**Comment by kroening** [comment]
Note that goto-convert will eventually move into the various language-specific directories, which may make some of this unnecessary.

There will be a base class in goto-programs, but that might end up being very basic.

File: src/goto-programs/goto_convert.cpp
Comment (tautschnig) [reply src/goto-programs/goto_convert.cpp]: I'm taking a look.

**Review by allredj — APPROVED** [review]
This seems to make sense as a whole. I was wondering if it would be better to pass a struct as argument to all these functions to avoid this kind of hurdle in the future but:
1) it touches too much code so I can't see a solution that would work well everywhere
2) after talking with @romainbrenguier we cannot think of another piece of data that should also be passed alongside the mode, so the proposed solution looks sensible to me.

File: src/goto-programs/goto_convert.cpp
Comment (tautschnig) [reply src/goto-programs/goto_convert.cpp]: Even using `symbol_table.lookup_ref(i.function)` fails, which in a way is good, because it means `symbol_table` and `ns` are in sync, but it also means that the current patch yields invalid memory accesses. The problem seems to be that the `function` field is not properly set. @peterschrammel had worked around this in 552b10080b63, but it may be necessary to dig deeper now.

Comment (romainbrenguier) [reply src/goto-programs/goto_convert.cpp]: @tautschnig I removed this lookup and just used `mode` coming from the calling function.

**Review by tautschnig — APPROVED** [review]

File: src/goto-programs/goto_convert.cpp
Comment (tautschnig) [reply src/goto-programs/goto_convert.cpp]: Ok, makes sense! I'll still invest some time to debug the reasons for the problem as it could be useful otherwise, but your approach is a very good one to unblock things.

**Comment by tautschnig** [comment]
@romainbrenguier Does this PR now make TG pass all tests? While I have no insight into this I'd like to get back to that state as soon as possible so that other PRs can be TG-checked before being merged.

File: src/goto-programs/goto_convert.cpp
Comment (tautschnig) [reply src/goto-programs/goto_convert.cpp]: Ok, `finish_gotos` will never be able to use `i.function`, because that field is only set later from `goto_convert_functionst::convert_function`. The chosen approach is probably the only correct one.

**Comment by romainbrenguier** [comment]
@tautschnig 
> Does this PR now make TG pass all tests? 

yes it does

**Review by peterschrammel — APPROVED** [review]

**Pull request merged**
